src/main/flight/pid.c

```diff
@@ -522,9 +526,11 @@ void pidController(const pidProfile_t *pidProfile, const rollAndPitchTrims_t *an
             gyroRateFiltered = dtermLowpassApplyFn((filter_t *) &dtermLowpass[axis], gyroRateFiltered);
             gyroRateFiltered = dtermLowpass2ApplyFn((filter_t *) &dtermLowpass2[axis], gyroRateFiltered);
 
-            const float rD = dynCd * MIN(getRcDeflectionAbs(axis) * relaxFactor, 1.0f) * currentPidSetpoint - gyroRateFiltered;    // cr - y
+            // no transition if relaxFactor == 0
+            const float transition = constrainf(getRcDeflectionAbs(axis) * relaxFactor, relaxFactor == 0, 1);
+            const float rD = dynCd * transition * currentPidSetpoint - gyroRateFiltered;
             // Divide rate change by deltaT to get differential (ie dr/dt)
-            float delta = (rD - previousRateError[axis]) / deltaT;
+            const float delta = (rD - previousRateError[axis]) / deltaT;
```

Comment:
This is the place where the other discontinuity happens. previousRateError was calculated based on dynCd of the last loop cycle. You can look at the equations equivalently as set point weight being constant but modifiedCurrentPidSetpoint being scaled in from 0 to transition. Then modifiedCurrentPidSetpoint is quadratic in stick from 0 to transition, then linear, and its derivative is directly given to D.

So clearly this causes a discontinuity in D at transition. I suspect what is actually intended here is to use the same current dynCd for both the old and the new currentPidSetpoint and give the pure derivative of currentPidSetpoint to D, weighted by current dynC.

Reply:
@joelucid - do you mean that, after this PR, when transition and weight are not zero, there remains a potential problem in the D calculation?  If transition was set to 0.01 by the user, the outcome would carry over the problem I noted above.  I can't see any benefit in flight to setting it so low :-) But even if set to 0.1 or 0.2 there is a sharp point reversal at centre in the D algorithm output.  The only solution to that is to use some kind of smooth function around centre.  Most pilots I know either use transition = 1, 0.5, or now 0.  And those who use 0.5 don't complain about mid stick problems, and even I barely noticed it when set to 0.01.

Anyway if you could clarify what you think needs to be done that would be great - thanks!

Reply:
OK great I understand it now I've seen your pull request:

https://github.com/betaflight/betaflight/pull/5509

Thank you!